File: electron/ipc/window.js
/**
 * Window IPC handlers — extracted from main.js.
 * Window management: minimize, maximize, close, new window, drag.
 */
const { ipcMain, BrowserWindow, nativeImage } = require('electron')
const path = require('path')
const fs = require('fs')

module.exports = function registerWindowHandlers(getMainWindow, PORT, appIcon) {
  // Idempotent registration — dev reloads may re-execute this module.
  const handles = ['window:minimize', 'window:maximize', 'window:unmaximize', 'window:close', 'window:isMaximized', 'window:toggleDevTools', 'window:newWindow']
  for (const channel of handles) {
    try { ipcMain.removeHandler(channel) } catch { /* ignore */ }
  }
  ipcMain.removeAllListeners('window:startDrag')

  function openNewWindow() {
    const win = new BrowserWindow({
      width: 1400, height: 900, minWidth: 800, minHeight: 600,
      title: 'Helix', icon: appIcon, frame: false, show: false, roundedCorners: 'off',
      webPreferences: {
        preload: path.join(__dirname, '..', 'preload.js'),
        contextIsolation: true, nodeIntegration: false, sandbox: false,
      },
      backgroundColor: '#FCFBF9',
    })
    const url = `http://localhost:${PORT}`
    win.loadURL(url)
    win.setMenuBarVisibility(false)
    win.once('ready-to-show', () => {
      win.maximize()
      win.show()
      win.focus()
    })
    if (process.platform === 'win32') {
      try {
        const winIcon = nativeImage.createFromPath(path.join(__dirname, '..', '..', 'public', 'Helix.ico'))
        if (!winIcon.isEmpty()) win.setIcon(winIcon)
      } catch (e) { console.log('Failed to set window icon:', e) }
    }
    win.on('maximize', () => win.webContents.send('window:maximized-changed', true))
    win.on('unmaximize', () => win.webContents.send('window:maximized-changed', false))
    return win
  }

  ipcMain.handle('window:minimize', (event) => {
    BrowserWindow.fromWebContents(event.sender)?.minimize()
  })

  ipcMain.handle('window:maximize', (event) => {
    BrowserWindow.fromWebContents(event.sender)?.maximize()
  })

  ipcMain.handle('window:unmaximize', (event) => {
    BrowserWindow.fromWebContents(event.sender)?.unmaximize()
  })

  ipcMain.handle('window:close', (event) => {
    BrowserWindow.fromWebContents(event.sender)?.close()
  })

  ipcMain.handle('window:isMaximized', (event) => {
    return BrowserWindow.fromWebContents(event.sender)?.isMaximized() ?? false
  })

  ipcMain.handle('window:toggleDevTools', (event) => {
    const win = BrowserWindow.fromWebContents(event.sender)
    if (win?.webContents.isDevToolsOpened()) {
      win.webContents.closeDevTools()
    } else {
      win?.webContents.openDevTools()
    }
  })

  ipcMain.handle('window:newWindow', () => {
    openNewWindow()
  })

  ipcMain.on('window:startDrag', () => {
    // No-op: handled by CSS -webkit-app-region: drag
  })

  return { openNewWindow }
}

File: src-tauri/src/window.rs
//! `window:*` commands — window management. Port of `electron/ipc/window.js`.

use crate::state::app_handle;
use tauri::Window;

#[tauri::command]
pub fn minimize(window: Window) {
    let _ = window.minimize();
}

#[tauri::command]
pub fn maximize(window: Window) {
    let _ = window.maximize();
}

#[tauri::command]
pub fn unmaximize(window: Window) {
    let _ = window.unmaximize();
}

#[tauri::command]
pub fn close(window: Window) {
    let _ = window.close();
}

#[tauri::command]
pub fn is_maximized(window: Window) -> bool {
    window.is_maximized().unwrap_or(false)
}

#[tauri::command]
pub fn toggle_devtools(window: Window) {
    // Devtools live on the webview, not the window. Use the first webview.
    if let Some(wv) = window.webviews().first() {
        if wv.is_devtools_open() {
            wv.close_devtools();
        } else {
            wv.open_devtools();
        }
    }
}

#[tauri::command]
pub fn new_window() {
    let handle = app_handle();
    let _ = tauri::WebviewWindowBuilder::new(
        handle,
        format!(
            "helix-win-{}",
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_millis())
                .unwrap_or(0)
        ),
        tauri::WebviewUrl::App("/".into()),
    )
    .title("Helix")
    .inner_size(1400.0, 900.0)
    .min_inner_size(800.0, 600.0)
    .decorations(false)
    .visible(false)
    .build()
    .and_then(|win| {
        let _ = win.maximize();
        let _ = win.show();
        let _ = win.set_focus();
        Ok(())
    });
}

#[tauri::command]
pub fn start_drag(window: Window) {
    // No-op: handled by CSS -webkit-app-region: drag (mirror electron behavior).
    let _ = window.start_dragging();
}
